core/src/main/java/org/elasticsearch/action/search/RemoteClusterConnection.java

```diff
@@ -0,0 +1,507 @@
+import org.elasticsearch.cluster.node.DiscoveryNodes;
+import org.elasticsearch.common.component.AbstractComponent;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.CancellableThreads;
+import org.elasticsearch.common.util.concurrent.AbstractRunnable;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.ConnectTransportException;
+import org.elasticsearch.transport.ConnectionProfile;
+import org.elasticsearch.transport.TcpTransport;
+import org.elasticsearch.transport.Transport;
+import org.elasticsearch.transport.TransportActionProxy;
+import org.elasticsearch.transport.TransportConnectionListener;
+import org.elasticsearch.transport.TransportException;
+import org.elasticsearch.transport.TransportRequest;
+import org.elasticsearch.transport.TransportRequestOptions;
+import org.elasticsearch.transport.TransportResponseHandler;
+import org.elasticsearch.transport.TransportService;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Predicate;
+
+/**
+ * Represents a connection to a single remote cluster. In contrast to a local cluster a remote cluster is not joined such that the
+ * current node is part of the cluster and it won't receive cluster state updates from the remote cluster. Remote clusters are also not
+ * fully connected with the current node. From a connection perspective a local cluster forms a bi-directional star network while in the
+ * remote case we only connect to a subset of the nodes in the cluster in an uni-directional fashion.
+ *
+ * This class also handles the discovery of nodes from the remote cluster. The initial list of seed nodes is only used to discover all nodes
+ * in the remote cluster and connects to all eligible nodes, for details see {@link RemoteClusterService#REMOTE_NODE_ATTRIBUTE}.
+ *
+ * In the case of a disconnection, this class will issue a re-connect task to establish at most
+ * {@link RemoteClusterService#REMOTE_CONNECTIONS_PER_CLUSTER} until either all eligible nodes are exhausted or the maximum number of
+ * connections per cluster has been reached.
+ */
+final class RemoteClusterConnection extends AbstractComponent implements TransportConnectionListener, Closeable {
+
+    private final TransportService transportService;
+    private final ConnectionProfile remoteProfile;
+    private final CopyOnWriteArrayList<DiscoveryNode> connectedNodes = new CopyOnWriteArrayList<>();
```

Comment:
@Maybe a `Collections.newSetFromMap(new ConcurrentHashMap<>())` would be more suited to the operations done on this collection (for example, avoiding the linear operations for `add` and `remove`) and the fact that `CopyOnWriteArrayList` is lock heavy?